aya/src/obj/btf/btf.rs

```diff
@@ -291,6 +360,212 @@ impl Btf {
             type_id: root_type_id,
         })
     }
+
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        // Safety: btf_header is POD
+        let mut buf = unsafe { bytes_of::<btf_header>(&self.header).to_vec() };
+        for t in self.types() {
+            let b = t.to_bytes();
+            buf.put(b.as_slice())
+        }
+        buf.put(self.strings.as_slice());
+        buf
+    }
+
+    pub(crate) fn fixup(
+        &mut self,
+        section_sizes: HashMap<String, u64>,
+        symbol_offsets: HashMap<String, u64>,
+    ) -> Result<(), BtfError> {
+        // FIXME: there is probably a more elegant way of doing operation in place
+        // for now, and to keep on the good side of the borrow checker, we'll create
+        // a new Vec and populate it as we go
+        let mut types = vec![];
+        for t in &self.types {
+            let kind = t.kind()?.unwrap_or_default();
+            // datasec sizes aren't set by llvm
+            // we need to fix them here before loading the btf to the kernel
+            match t {
+                BtfType::Ptr(mut ty) => {
+                    // Rust emits names for pointer types, which the kernel doesn't like
+                    // While I figure out if this needs fixing in the Kernel or LLVM, we'll
+                    // do a fixup here
+                    ty.name_off = 0;
+                    types.push(BtfType::Ptr(ty));
+                }
+                BtfType::DataSec(mut ty, data) => {
+                    // Start DataSec Fixups
+                    let sec_name = self.type_name(t)?.ok_or(BtfError::InvalidTypeInfo)?;
+                    let name = sec_name.to_string();
+                    // There are cases when the compiler does indeed populate the
```

Comment:
Do you have a test case that repros when LLVM sets the size vs when it doesn't? I'm happy to fix this in LLVM if it's a bug

Reply:
It's expected behaviour. See https://reviews.llvm.org/D59441

> At the time of debug info emission, the data section
size is unknown, so the btf_type.size = 0 for
BTF_KIND_DATASEC.

I don't remember if I saw a comment in libbpf or LLVM that triggered me to add this conditional